### PR TITLE
Adding Serializer component with interface and simple implementation

### PR DESCRIPTION
**Pull request by msmakouz**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌ 
| New feature?  | ✔️

Added component with serializer interface and some simple default implementations with wrappers over standard PHP functions. In a real application, it is supposed to use the integration of more powerful external serializer packages. For example, symfony/serializer, our marshaller.

To add an external serializer, need to create a wrapper over the serializer, implement the `Spiral\Serializer\SerializerInterface` interface, and register it in serializer config  or in the `Spiral\Serializer\SerializerRegistryInterface` before using `SerializerInterface` or `SerializerManager` for the first time. For example, in the init or boot method of the bootloader.

Example with config:
```php
// file app/config/serializer.php
<?php

declare(strict_types=1);

use Spiral\Core\Container\Autowire;

return [
    'default' => env('DEFAULT_SERIALIZER_FORMAT', 'json'),
    'serializers' => [
        'json' => JsonSerializer::class,
        'yaml' => new YamlSerializer(),
        'custom' => new Autowire(CustomSerializer::class),
    ],
];
```
Example with SerializerRegistry:
```php
final class SerializerBootloader extends Bootloader
{
    private function init(SerializerRegistryInterface $serializers): void
    {
        $serializers->register('json', new JsonSerializer());
        $serializers->register('yaml', new YamlSerializer());

        $serializers->register('custom', new CallbackSerializer(
            serialize: fn(mixed $payload) => \json_encode($payload),
            unserialize: fn(string|\Stringable $payload, string|object|null $type = null) => \json_decode($payload, true),
        ));
    }
}

```

### CODE REVIEW

File: src/Serializer/src/Bootloader/SerializerBootloader.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        return new SerializerRegistry(\array_map([$this, 'wire'], $config->getSerializers()));
+    }
+
+    private function initConfig(Environment $env): void
+    {
+        $this->config->setDefaults(SerializerConfig::CONFIG, [
+            'default' => $env->get('DEFAULT_SERIALIZER_FORMAT', SerializerConfig::DEFAULT_SERIALIZER),
+            'serializers' => [
+                'json' => new JsonSerializer(),
+                'serialize' => new PhpSerializer(),
```
Comment (butschster) [inline src/Serializer/src/Bootloader/SerializerBootloader.php:62]: Maybe `php` or `serializer`?

File: src/Serializer/src/Bootloader/SerializerBootloader.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+use Spiral\Serializer\SerializerInterface;
+use Spiral\Serializer\SerializerRegistry;
+use Spiral\Serializer\SerializerManager;
+use Spiral\Serializer\SerializerRegistryInterface;
+
+final class SerializerBootloader extends Bootloader
+{
+    protected const SINGLETONS = [
+        SerializerManager::class => [self::class, 'initSerializerManager'],
+        SerializerRegistryInterface::class => [self::class, 'initSerializerRegistry'],
```
Comment (butschster) [inline src/Serializer/src/Bootloader/SerializerBootloader.php:24]: Maybe add additional singleton for SerializerInterface?

```
SerializerInterface::class => SerializerManager::()/
```

**Review by butschster — CHANGES_REQUESTED** [review]

File: src/Serializer/src/Bootloader/SerializerBootloader.php (hunk 2 above)
Comment (msmakouz) [reply src/Serializer/src/Bootloader/SerializerBootloader.php:24]: Added

File: src/Serializer/src/SerializerManager.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        return $this->getSerializer($format ?? $this->defaultFormat)->serialize($payload, $format);
+    }
+
+    public function unserialize(
+        string|\Stringable $payload,
+        string|object|null $type = null,
+        ?string $format = null
+    ): mixed {
+        return $this->getSerializer($format ?? $this->defaultFormat)->unserialize($payload, $type, $format);
```
Comment (roxblnfk) [inline src/Serializer/src/SerializerManager.php:30]:
```suggestion
        return $this->getSerializer($format ?? $this->defaultFormat)->unserialize($payload, $type);
```
Should `$format` be here in the `unserialize` calling?

File: src/Serializer/src/SerializerManager.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    public function getSerializer(string $format = null): SerializerInterface
+    {
+        return $this->serializers->get($format ?? $this->defaultFormat);
+    }
+
+    public function serialize(mixed $payload, ?string $format = null): string|\Stringable
+    {
+        return $this->getSerializer($format ?? $this->defaultFormat)->serialize($payload, $format);
```
Comment (roxblnfk) [inline src/Serializer/src/SerializerManager.php:22]:
```suggestion
        return $this->getSerializer($format ?? $this->defaultFormat)->serialize($payload);
```
The same

File: src/Serializer/src/Serializer/PhpSerializer.php [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+final class PhpSerializer implements SerializerInterface
+{
+    public function serialize(mixed $payload): string|\Stringable
+    {
+        return serialize($payload);
+    }
+
+    public function unserialize(\Stringable|string $payload, object|string|null $type = null): mixed
+    {
+        return unserialize((string) $payload);
```
Comment (roxblnfk) [inline src/Serializer/src/Serializer/PhpSerializer.php:18]: If the `$type` argumend has been passed should we process it?
For example:
$type is class name -> we should check the unserialize result is instance of $type;
$type is object -> throw an exception about unsupported argument type OR hydrate that object

File: src/Serializer/src/SerializerManager.php (hunk 3 above)
Comment (msmakouz) [reply src/Serializer/src/SerializerManager.php]: This is a mistake, removed

File: src/Serializer/src/SerializerManager.php (hunk 4 above)
Comment (msmakouz) [reply src/Serializer/src/SerializerManager.php]: Removed

File: src/Serializer/src/Serializer/PhpSerializer.php (hunk 5 above)
Comment (msmakouz) [reply src/Serializer/src/Serializer/PhpSerializer.php]: Added exception.

File: src/Serializer/src/Serializer/JsonSerializer.php [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        return \json_encode($payload, JSON_THROW_ON_ERROR);
+    }
+
+    /**
+     * @throws \JsonException
+     */
+    public function unserialize(\Stringable|string $payload, object|string|null $type = null): mixed
+    {
+        return \json_decode((string) $payload, true, 512, JSON_THROW_ON_ERROR);
+    }
```
[the remark this answers is not in the document]
Comment (msmakouz) [reply src/Serializer/src/Serializer/JsonSerializer.php]: Added validation with type checking

**Review by butschster — COMMENTED** [review]
Fix exception messages

**Review by butschster — CHANGES_REQUESTED** [review]
Fix exception message

File: src/Serializer/src/Serializer/JsonSerializer.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     * @throws \JsonException
+     */
+    public function unserialize(\Stringable|string $payload, object|string|null $type = null): mixed
+    {
+        if ($type !== null) {
+            throw new UnserializeException(
+                \sprintf('Serializer %s does not support data hydration to an objects', self::class)
```
Comment (butschster) [inline src/Serializer/src/Serializer/JsonSerializer.php:27]: Replace with `'Serializer [%s] does not support data hydration to an object.'`

File: src/Serializer/src/Serializer/PhpSerializer.php [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+        if (\is_object($type)) {
+            $type = $type::class;
+        }
+
+        $result = unserialize((string) $payload, ['allowed_classes' => [$type]]);
+
+        if (!$result instanceof $type) {
+            throw new UnserializeException(\sprintf(
+                'Data received after unserializing must be of type: %s, received %s.',
```
Comment (butschster) [inline src/Serializer/src/Serializer/PhpSerializer.php:31]: Replace with `'Data received after unserializing must be of type: [%s], received [%s].'`

**Review by butschster — CHANGES_REQUESTED** [review]

File: src/Serializer/src/Serializer/JsonSerializer.php [hunk 9]
```diff
@@ -21,6 +22,12 @@ public function serialize(mixed $payload): string|\Stringable
      */
     public function unserialize(\Stringable|string $payload, object|string|null $type = null): mixed
     {
+        if ($type !== null) {
+            throw new UnserializeException(
```
Comment (roxblnfk) [inline src/Serializer/src/Serializer/JsonSerializer.php:26]: What about InvalidArgumentException that extends UnserializeException

File: src/Serializer/src/Serializer/PhpSerializer.php (hunk 8 above)
Comment (msmakouz) [reply src/Serializer/src/Serializer/PhpSerializer.php]: Done

File: src/Serializer/src/Serializer/JsonSerializer.php (hunk 9 above)
Comment (msmakouz) [reply src/Serializer/src/Serializer/JsonSerializer.php:26]: Done.

**Review by butschster — APPROVED** [review]

**Pull request merged**